lib/puppet/pops/loader/loader.rb

```diff
@@ -137,43 +137,51 @@ def initialize(typed_name, value, origin)
   # A name/type combination that can be used as a compound hash key
   #
   class TypedName
+    DOUBLE_COLON = '::'
+
+    attr_reader :hash
     attr_reader :type
     attr_reader :name
     attr_reader :name_parts
-
-    # True if name is qualified (more than a single segment)
-    attr_reader :qualified
+    attr_reader :compound_name
 
     def initialize(type, name)
       @type = type
       # relativize the name (get rid of leading ::), and make the split string available
-      @name_parts = name.to_s.split(/::/)
-      @name_parts.shift if name_parts[0].empty?
-      @name = name_parts.join('::')
-      @qualified = name_parts.size > 1
-      # precompute hash - the name is frozen, so this is safe to do
-      @hash = [self.class, type, @name].hash
-
-      # Not allowed to have numeric names - 0, 010, 0x10, 1.2 etc
-      if Puppet::Pops::Utils.is_numeric?(@name)
-        raise ArgumentError, "Illegal attempt to use a numeric name '#{name}' at #{origin_label(origin)}."
+      parts = name.to_s.split(DOUBLE_COLON)
+      if parts[0].empty?
+        parts.shift
+        @name = name[2..-1]
+      else
+        @name = name
       end
+      @name_parts = parts
 
-      freeze()
-    end
+      # Not allowed to have numeric names - 0, 010, 0x10, 1.2 etc.
+      # At this point it is known that @name is a relativized string so it's
+      # safe to do a direct regexp match
+      unless Puppet::Pops::Patterns::NUMERIC.match(@name).nil?
```

Comment:
I wonder if the loader/typed-name really needs to care about numeric names - I don't think there is any ambiguity in the loaders. Names of things should be checked at the source (i.e. "illegal function name", "illegal class name" etc.) as there are far fewer of those than instances of TypedName.

That was a long way of saying: I think we can simply skip this assertion.

(sorry, I first commented on the commit - does not show up in the conversation).